### ISSUE TITLE
Intended use for middleware_names

### ISSUE THREAD

**Issue by callumj** [DataDog/dd-trace-rb#1261]
Hi there,

When using `middleware_names` with Rack & Rails should we expect to see a dedicated span generated for each middleware or is it intended to just try to capture the most recent middleware in the chain? So if we wanted to trace each middleware that would be custom?

**Comment by callumj** [comment]
Semi related: shouldn't request queuing be computed when `ActionDispatch::Routing::RouteSet` is hit (the very first Rails middleware) to get a better value on when the request first hits the application?

**Comment by marcotc** [comment]
Hi @callumj, thank you for reaching out.

> When using middleware_names with Rack & Rails should we expect to see a dedicated span generated for each middleware or is it intended to just try to capture the most recent middleware in the chain? So if we wanted to trace each middleware that would be custom?

This code predates me by quite a bit, but I just too a look now and I believe that we are indeed only recording one middleware name, as I don't see any code changes driven by the `middleware_names` that can cause additional `spans` to be generated.

Our documentation says that this option makes our `rack` instrumentation _"use the middleware classes as the resource names for `rack` spans."_. This is interesting, given that there's very likely only a single Rack span for a web request. So it's technically achieving what is documented, but this doesn't seem of much value, in my opinion.

I'll see if I can dig a little bit of this history behind the implementation of this feature, and work on actually capturing all middleware spans if we can't find a good reason to maintain the current implementation.

> Semi related: shouldn't request queuing be computed when ActionDispatch::Routing::RouteSet is hit (the very first Rails middleware) to get a better value on when the request first hits the application?

The way the data structure for traces work, to capture the time spent by a caller entity (a reverse proxy in this case), we have to measure the total time spent on that caller. If you need to know the "self" time of the caller before the first child starts, you'd have to subtract the start time of the caller span from the start time of the first child span _after_ the data structure has been recorded.

In short, the answer you are looking can be calculated from the data structure stored today, but we don't have an easy way to display this information in our UI today.

For context, this is not the first time such request has come to our attention. I'll follow up with our UI teams to see if we have any blockers to surface this information.

**Comment by marcotc** [comment]
I have an answer for the intended use of `middleware_names`: https://github.com/DataDog/dd-trace-rb/pull/285

Here's the gist of it:
> Let's say you have an AuthenticationMiddleware. In a Rails application context, an unauthorized request would never hit the Rails stack and it would surface as something like aGET 401 resource name.
By enabling the middleware_names option will you get a AuthenticationMiddleware#GET resource name instead.

It always records the last middleware to be executed, which should match the middleware responsible for generating the response object.

If the Rails integration is also enabled, we actually override the Rack span resource name with the Rails controller name, and this will take precedence over `middleware_names`. `middleware_names` can still act as a fail-safe for the Rack span in case the request does not reach the controller.

So recording every single middleware is not what the feature is meant for. I'm assuming you are interested in recording every middleware, is that right?

**Comment by marcotc** [comment]
I've updated our documentation with the findings above: https://github.com/DataDog/dd-trace-rb/pull/1262

The existing description did not describe precisely the behaviour intended for that option.

**Comment by callumj** [comment]
Thank you @marcotc, that does seem to line up with what I am reading. It doesn't seem like DD APM can provide a way to monitor the diff between `SPAN_HTTP_SERVER_QUEUE` & `SPAN_REQUEST` spans to build an effective monitor for slow request queueing so for now I am opting to instrument the difference using statsd:

```ruby
  request_queue_tags = Datadog.configuration.tags.map { |k, v| "#{k}:#{v}" }
  Datadog::Pipeline.before_flush(
    Datadog::Pipeline::SpanProcessor.new do |span|
      puts span.resource unless span.resource.includes?("TRUNCATE")
      case span.name
      when Datadog::Contrib::Rack::Ext::SPAN_HTTP_SERVER_QUEUE
        Thread.current[:datadog_apm_http_server_queue_start] = span.start_time
      when Datadog::Contrib::Rack::Ext::SPAN_REQUEST
        next unless Thread.current[:datadog_apm_http_server_queue_start]

        duration = Thread.current[:datadog_apm_http_server_queue_start] - span.start_time
        DOGSTATSD.timing(
          'monorail.request_queue_time',
          duration,
          tags: request_queue_tags,
        )
      end
    end
  )
```



Rack middleware instrumentation would be great, for now I am doing this with some custom patching:

```ruby
  def self.patch_rack_middlewares
    return if @middleware_patched

    target_middleware = get_upper_middleware(Rails.application)

    while target_middleware
      this_middleware = target_middleware
      target_middleware = get_upper_middleware(target_middleware)
      next if IGNORED_MIDDLEWARE.include?(target_middleware.class)
      next unless this_middleware.respond_to?(:call)
      # double check we didn't already patch this
      # or something else did (to avoid infinite stack calls)
      next if this_middleware.respond_to?(:__call_before_middleware_patch)

      this_middleware.singleton_class.class_eval do
        alias_method :__call_before_middleware_patch, :call

        def call(env)
          Datadog.tracer.trace('rack.middleware', apm_rack_middleware_span_options) do |span|
            __call_before_middleware_patch(env)
          end
        end

        def apm_rack_middleware_span_options
          @span_options ||= {
            span_type: Datadog::Ext::HTTP::TYPE_INBOUND,
            resource: self.class.to_s,
          }
        end
      end
    end
  ensure
    @middleware_patched = true
  end

  def self.get_upper_middleware(middleware)
    return unless middleware.instance_variable_defined?('@app')

    middleware.instance_variable_get('@app')
  end
```

**Comment by callumj** [comment]
It looks like the middleware wrapping doesn't work out of the box because `Datadog::Contrib::Rack::TraceMiddleware` resets the context on finish which will orphan all the other traces started in upstream middlewares: https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/contrib/rack/middlewares.rb#L122

**Comment by callumj** [comment]
It seems this attempt at middleware tracing is pretty unreliable. My traces rarely have all these middleware traces and it appears to may be because the root `trace` isn't present in the context. I looked at calls to `Context.get` and the `@trace` would be `nil`.

**Comment by callumj** [comment]
Here is an output from debug, it appears that after the first run the middleware spans become orphaned from the `rack.request` spans - I even tried passing in `child_of: env["datadog.rack_request_span"]:

```
Started GET "/health" for 127.0.0.1 at 2020-11-26 01:19:39 +0000
[ddtrace] (<HIDDEN>) Writing 6 spans (enabled: true)
[ddtrace] (<HIDDEN>) [
 Name: rack.middleware
Span ID: 4299606000550878393
Parent ID: 0
Trace ID: 3854317489117373906
Type: custom
Service: myapp-ddtrace
Resource: ExampleApp::Tracking::Rack::Middleware
Error: 0
Start: 1606353692866773504
End: 1606353692875225088
Duration: 0.008450641000081305
Allocations: 5863
Tags: [
   runtime-id => eb42997a-0ff7-441a-972c-212cfaaf6ed0,
   chef.role => Unknown,
   env => Unknown,
   service => myapp-ddtrace,
   version => Unknown,
   application => myapp-ddtrace]
Metrics: [
   _dd.agent_psr => 1.0,
   system.pid => 12264.0,
   _sampling_priority_v1 => 1.0],
 
 Name: rack.middleware
Span ID: 612883332299987801
Parent ID: 4299606000550878393
Trace ID: 3854317489117373906
Type: custom
Service: myapp-ddtrace
Resource: PerformanceMetricsMiddleware
Error: 0
Start: 1606353692866843648
End: 1606353692875202816
Duration: 0.008364338000092175
Allocations: 5809
Tags: [
   chef.role => Unknown,
   env => Unknown,
   service => myapp-ddtrace,
   version => Unknown,
   application => myapp-ddtrace]
Metrics: [ ],
 
 Name: rack.middleware
Span ID: 1881975062314124639
Parent ID: 612883332299987801
Trace ID: 3854317489117373906
Type: custom
Service: myapp-ddtrace
Resource: RackStats::Middleware
Error: 0
Start: 1606353692867015936
End: 1606353692874044416
Duration: 0.007031402999928105
Allocations: 4531
Tags: [
   chef.role => Unknown,
   env => Unknown,
   service => myapp-ddtrace,
   version => Unknown,
   application => myapp-ddtrace]
Metrics: [ ],
 
 Name: rack.middleware
Span ID: 3102763057142708425
Parent ID: 1881975062314124639
Trace ID: 3854317489117373906
Type: custom
Service: myapp-ddtrace
Resource: ServiceRequestStore::Middleware
Error: 0
Start: 1606353692871889152
End: 1606353692873444608
Duration: 0.0015484880000258272
Allocations: 1226
Tags: [
   chef.role => Unknown,
   env => Unknown,
   service => myapp-ddtrace,
   version => Unknown,
   application => myapp-ddtrace]
Metrics: [ ],
 
 Name: rack.middleware
Span ID: 823152941964054225
Parent ID: 3102763057142708425
Trace ID: 3854317489117373906
Type: custom
Service: myapp-ddtrace
Resource: ServiceRequestStore::Middleware
Error: 0
Start: 1606353692871939840
End: 1606353692873424384
Duration: 0.0014887340000768745
Allocations: 1202
Tags: [
   chef.role => Unknown,
   env => Unknown,
   service => myapp-ddtrace,
   version => Unknown,
   application => myapp-ddtrace]
Metrics: [ ],
 
 Name: rack.middleware
Span ID: 1286258749042733719
Parent ID: 823152941964054225
Trace ID: 3854317489117373906
Type: custom
Service: myapp-ddtrace
Resource: GoogleBotAware
Error: 0
Start: 1606353692872005632
End: 1606353692872829952
Duration: 0.0008308869998927548
Allocations: 905
Tags: [
   chef.role => Unknown,
   env => Unknown,
   service => myapp-ddtrace,
   version => Unknown,
   application => myapp-ddtrace]
Metrics: [ ]]

[ddtrace] (<HIDDEN>) Writing 1 spans (enabled: true)
[ddtrace] (<HIDDEN>) [
 Name: rack.request
Span ID: 2043053818134663127
Parent ID: 0
Trace ID: 3669813217685611195
Type: web
Service: myapp-ddtrace
Resource: HealthApp
Error: 0
Start: 1606353692866625792
End: 1606353692876225536
Duration: 0.009602914999959467
Allocations: 7061
Tags: [
   runtime-id => eb42997a-0ff7-441a-972c-212cfaaf6ed0,
   chef.role => Unknown,
   env => Unknown,
   service => myapp-ddtrace,
   version => Unknown,
   application => myapp-ddtrace,
   language => ruby,
   http.method => GET,
   http.url => http://www.example.com/health,
   http.base_url => http://www.example.com,
   http.status_code => 200,
   http.request.headers.user_agent => curl/7.35.0,
   http.request.headers.accept => */*,
   http.request.headers.host => www.example.com,
   http.response.headers.content_type => text/plain]
Metrics: [
   _dd.agent_psr => 1.0,
   system.pid => 12264.0,
   _dd.measured => 1.0,
   _sampling_priority_v1 => 1.0]]
```

**Comment by delner** [comment]
@callumj Is the general goal of this to instrument all middleware? As a feature request? Just trying to categorize this for our own consideration.

**Comment by delner** [comment]
I think this is more or less a duplicate of https://github.com/DataDog/dd-trace-rb/issues/368

Going to close this one in favor of that one. Please feel free to comment if you think otherwise.
